### PR TITLE
[9.0][MIG] account_refund_original > account_invoice_refund_link

### PR DESCRIPTION
**Pull request by antespi**
- Rename addon to `account_invoice_refund_link`
- OCA Guidelines complaince 
- Test added
- Rename field `origin_invoices_ids` to `origin_invoice_ids`
- Rename field `refund_invoices_description` to `refund_reason`
- Show refund invoices created in a normal invoice
- Link invoice lines between refund an original invoice

### CODE REVIEW

File: account_invoice_refund_link/README.rst [hunk 1]
```diff
@@ -0,0 +1,65 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=================================
+Link refund invoice with original
```
Comment (pedrobaeza) [inline account_invoice_refund_link/README.rst]: Link the refund invoice with its source invoice

File: account_invoice_refund_link/README.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+This module links customer and supplier refunds with the invoice that originate
+them (and invoice lines).
+
+Usage
+=====
+
+In invoice form there is a new 'Refund' tab:
+
+For refund invoices:
+- Show the refund reason
```
Comment (pedrobaeza) [inline account_invoice_refund_link/README.rst]: s/Show/It shows

File: account_invoice_refund_link/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
+# Copyright 2004-2011 Pexego Sistemas Informáticos. (http://pexego.es)
+# Copyright 2014 Pedro M. Baeza <pedro.baeza@serviciosbaeza.com>
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Link refund invoice with original",
+    "summary": "Link refund invoice with its original invoice",
+    "version": "9.0.1.0.0",
+    "author": "Spanish Localization Team, "
```
Comment (pedrobaeza) [inline account_invoice_refund_link/__openerp__.py]: Unfold here all the companies: `Pexego` and remove `Spanish Localization Team`.

File: account_invoice_refund_link/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+              "Tecnativa, "
+              "Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/l10n-spain",
```
Comment (pedrobaeza) [inline account_invoice_refund_link/__openerp__.py]: Change this URL

File: account_invoice_refund_link/hooks.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            ('type', 'in', ('out_refund', 'in_refund')),
+        ])
+        _logger.info(
+            "Linking %d refund invoices", len(refunds))
+        for refund in refunds:
+            original = _invoice_match(env, refund)
+            if original:
+                refund.write({
+                    'origin_invoice_ids': [(6, 0, original.ids)],
+                    'refund_reason': 'Auto',
```
Comment (pedrobaeza) [inline account_invoice_refund_link/hooks.py]: Translate this. Can't this be deducted?

File: account_invoice_refund_link/hooks.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                })
+                idx = 0
+                for line in original:
```
Comment (pedrobaeza) [inline account_invoice_refund_link/hooks.py]: `for idx, line in enumerate(original.invoice_line_ids):`

File: account_invoice_refund_link/hooks.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                    refund.invoice_line_ids[idx].write({
```
Comment (pedrobaeza) [inline account_invoice_refund_link/hooks.py]: Protect this, because some lines might be removed

File: account_invoice_refund_link/i18n/account_invoice_refund_link.pot [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# Translation of OpenERP Server.
```
Comment (pedrobaeza) [inline account_invoice_refund_link/i18n/account_invoice_refund_link.pot]: Remove POT file

File: account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp.addons.account_invoice_refund_link.hooks import post_init_hook
+
+
+def migrate(cr, version):
+    # Link all refund invoices to its original invoices
+    post_init_hook(cr, None)
```
Comment (pedrobaeza) [inline account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py]: This shouldn't be done in the migration script

File: account_invoice_refund_link/migrations/9.0.1.0.0/pre-migrate.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openupgradelib import openupgrade
+
+# Rename field: origin_invoices_ids -> origin_invoice_ids
```
Comment (pedrobaeza) [inline account_invoice_refund_link/migrations/9.0.1.0.0/pre-migrate.py]: There's no need of this comment. The dictionary is self-explanatory

File: account_invoice_refund_link/models/account_invoice.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    refund_reason = fields.Text(string="Refund reason")
+    origin_invoice_ids = fields.Many2many(
+        comodel_name='account.invoice', column1='refund_invoice_id',
+        column2='original_invoice_id', relation='account_invoice_refunds_rel',
+        string="Original invoice", readonly=True,
+        help="Links to original invoice which is referred this refund invoice")
+    refund_invoice_ids = fields.Many2many(
+        comodel_name='account.invoice', column1='original_invoice_id',
+        column2='refund_invoice_id', relation='account_invoice_refunds_rel',
+        string="Refund invoice",  readonly=True,
```
Comment (pedrobaeza) [inline account_invoice_refund_link/models/account_invoice.py]: This should be plural.

File: account_invoice_refund_link/models/account_invoice_line.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+# Copyright 2016 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    origin_line_ids = fields.Many2many(
```
Comment (pedrobaeza) [inline account_invoice_refund_link/models/account_invoice_line.py]: You can make also the inverse here

File: account_invoice_refund_link/views/account_invoice_view.xml [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+                        <field name="number" string="Number"/>
+                        <field name="reference" string="Reference"/>
+                        <field name="origin" string="Source"/>
+                        <field name="amount_total_signed" string="Total"/>
+                        <field name="residual_signed" string="Due"/>
+                        <field name="state" string="Status"/>
+                    </tree>
+                </field>
+            </page>
+            <page string="Refunds"
```
Comment (pedrobaeza) [inline account_invoice_refund_link/views/account_invoice_view.xml]: Why 2 pages with the same name? Use groups with attrs instead

File: account_invoice_refund_link/wizard/account_invoice_refund.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            return result
+        # An example of result['domain'] computed by the parent wizard is:
+        # [('type', '=', 'out_refund'), ('id', 'in', [43L, 44L])]
+        # The created refund invoice is the first invoice in the
+        # ('id', 'in', ...) tupla
+        created_inv = [x[2] for x in result['domain']
+                       if x[0] == 'id' and x[1] == 'in']
+        if created_inv and created_inv[0]:
+            description = self[0].description or ''
+            refund_inv_id = created_inv[0][0]
```
Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py]: Is this correct or should be `created_inv[0]`?

File: account_invoice_refund_link/wizard/account_invoice_refund.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+            refund = self.env['account.invoice'].browse(refund_inv_id)
+            refund.write({
+                'origin_invoice_ids': [(6, 0, active_ids)],
```
Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py]: Why are you matching all the possible source invoices?

File: account_invoice_refund_link/wizard/account_invoice_refund.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    def compute_refund(self, mode='refund'):
+        result = super(AccountInvoiceRefund, self).compute_refund(mode)
+        active_ids = self.env.context.get('active_ids')
+        if not active_ids:
+            return result
+        # An example of result['domain'] computed by the parent wizard is:
+        # [('type', '=', 'out_refund'), ('id', 'in', [43L, 44L])]
+        # The created refund invoice is the first invoice in the
+        # ('id', 'in', ...) tupla
+        created_inv = [x[2] for x in result['domain']
```
Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py:23]: This theoretically can be more than one invoice

**Comment by pedrobaeza** [comment]
Check Travis result

File: account_invoice_refund_link/hooks.py (hunk 5 above)
Comment (antespi) [inline account_invoice_refund_link/hooks.py]: Ok for translation here
I don't think this could be deduced automatically, any suggestions?

File: account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py (hunk 9 above)
Comment (antespi) [inline account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py]: Why?, if this version links all invoices when installing, it seems logical to do the same when upgrading to this version, isn't it?

File: account_invoice_refund_link/wizard/account_invoice_refund.py (hunk 14 above)
Comment (antespi) [inline account_invoice_refund_link/wizard/account_invoice_refund.py]: Yes, created_inv[0] is a list of ids

File: account_invoice_refund_link/hooks.py (hunk 5 above)
Comment (pedrobaeza) [inline account_invoice_refund_link/hooks.py]: Isn't the text put on name field?

File: account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py (hunk 9 above)
Comment (pedrobaeza) [inline account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py]: No, migrating is one thing, and hooks are another. If you want that on a migrated database, then call manually the hook.

File: account_invoice_refund_link/wizard/account_invoice_refund.py (hunk 16 above)
Comment (antespi) [inline account_invoice_refund_link/wizard/account_invoice_refund.py:23]: Yes, but this addon considers only the first one

Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py:23]: Then you should consider all, not only the first one. Here you can do it.

File: account_invoice_refund_link/wizard/account_invoice_refund.py (hunk 15 above)
Comment (antespi) [inline account_invoice_refund_link/wizard/account_invoice_refund.py]: Done, only get the first one

File: account_invoice_refund_link/views/account_invoice_view.xml (hunk 13 above)
Comment (antespi) [inline account_invoice_refund_link/views/account_invoice_view.xml]: Done

File: account_invoice_refund_link/models/account_invoice_line.py (hunk 12 above)
Comment (antespi) [inline account_invoice_refund_link/models/account_invoice_line.py]: Done

File: account_invoice_refund_link/models/account_invoice.py (hunk 11 above)
Comment (antespi) [inline account_invoice_refund_link/models/account_invoice.py]: Done

File: account_invoice_refund_link/migrations/9.0.1.0.0/pre-migrate.py (hunk 10 above)
Comment (antespi) [inline account_invoice_refund_link/migrations/9.0.1.0.0/pre-migrate.py]: Done

File: account_invoice_refund_link/i18n/account_invoice_refund_link.pot (hunk 8 above)
Comment (antespi) [inline account_invoice_refund_link/i18n/account_invoice_refund_link.pot]: Done

File: account_invoice_refund_link/hooks.py (hunk 7 above)
Comment (antespi) [inline account_invoice_refund_link/hooks.py]: Done, I've wrote a model method in order to respect DRY principle

File: account_invoice_refund_link/hooks.py (hunk 6 above)
Comment (antespi) [inline account_invoice_refund_link/hooks.py]: Thanks, good point!

File: account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py (hunk 9 above)
Comment (antespi) [inline account_invoice_refund_link/migrations/9.0.1.0.0/post-migrate.py]: Ok, done and squashed

**Comment by pedrobaeza** [comment]
Don't comment until everything is done. It's very hard to follow otherwise.

File: account_invoice_refund_link/models/account_invoice.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    @api.multi
+    def match_origin_lines(self, origin_inv):
+        for idx, line in enumerate(origin_inv.invoice_line_ids):
+            try:
+                # Protect this write, maybe refund invoice doesn't
+                # have the same lines than original one
+                self.invoice_line_ids[idx].write({
+                    'origin_line_ids': [(6, 0, line.ids)],
+                })
+            except:
```
Comment (pedrobaeza) [inline account_invoice_refund_link/models/account_invoice.py]: Put here and in the following line `# pragma: no cover` for avoiding a drop in coveralls without reason

**Comment by pedrobaeza** [comment]
Please test the hook this way:
- Create an invoice.
- Create a refund from this invoice.
- Delete manually the link between both.
- Call the hook.
- Check that the link has been reestablished.

**Comment by antespi** [comment]
Test added and squashed

File: account_invoice_refund_link/wizard/account_invoice_refund.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+
+class AccountInvoiceRefund(models.TransientModel):
+    _inherit = "account.invoice.refund"
+
+    @api.multi
+    def compute_refund(self, mode='refund'):
+        result = super(AccountInvoiceRefund, self).compute_refund(mode)
+        active_ids = self.env.context.get('active_ids')
+        if not active_ids:
+            return result
```
Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py:18]: Put here also `# pragma: nocover`

File: account_invoice_refund_link/wizard/account_invoice_refund.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+        # An example of result['domain'] computed by the parent wizard is:
+        # [('type', '=', 'out_refund'), ('id', 'in', [43L, 44L])]
+        # The created refund invoice is the first invoice in the
+        # ('id', 'in', ...) tupla
+        created_inv = [x[2] for x in result['domain']
+                       if x[0] == 'id' and x[1] == 'in']
+        if mode == 'modify':
+            # Remove pairs ids, because they are new draft invoices
+            del created_inv[1::2]
```
Comment (pedrobaeza) [inline account_invoice_refund_link/wizard/account_invoice_refund.py:27]: There's no test for this case. Please make one.

**Comment by antespi** [comment]
Test improved, bug fixed and squashed

**Comment by antespi** [comment]
Please add NeedsReview label to this PR, thanks.

File: account_invoice_refund_link/README.rst [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+This module links customer and supplier refunds with the invoice that originate
+them (and invoice lines).
+
+Usage
+=====
+
+In invoice form there is a new 'Refund' tab:
+
+For refund invoices:
+- It shows the refund reason
```
Comment (elicoidal) [inline account_invoice_refund_link/README.rst:19]: add empty line before list

File: account_invoice_refund_link/models/account_invoice.py [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    refund_reason = fields.Text(string="Refund reason")
+    origin_invoice_ids = fields.Many2many(
+        comodel_name='account.invoice', column1='refund_invoice_id',
+        column2='original_invoice_id', relation='account_invoice_refunds_rel',
+        string="Original invoice", readonly=True,
+        help="Links to original invoice which is referred this refund invoice")
```
Comment (elicoidal) [inline account_invoice_refund_link/models/account_invoice.py]: Original invoice to which this refund invoice is referred to

File: account_invoice_refund_link/models/account_invoice.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    refund_invoice_ids = fields.Many2many(
+        comodel_name='account.invoice', column1='original_invoice_id',
+        column2='refund_invoice_id', relation='account_invoice_refunds_rel',
+        string="Refund invoices",  readonly=True,
+        help="Links to refund invoices related with this invoice")
```
Comment (elicoidal) [inline account_invoice_refund_link/models/account_invoice.py]: Original refund to which this invoice is referred to

File: account_invoice_refund_link/models/account_invoice_line.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+        comodel_name='account.invoice.line', column1='refund_line_id',
+        column2='original_line_id', string="Original invoice line",
+        relation='account_invoice_line_refunds_rel',
+        help="Links to original invoice line which is referred by current "
```
Comment (elicoidal) [inline account_invoice_refund_link/models/account_invoice_line.py]: check previous messages (I would even simplify and make it like "Original invoice related to this refund")

**Comment by antespi** [comment]
@elicoidal Fixes done, thanks

**Comment by elicoidal** [comment]
:+1:

**Comment by antespi** [comment]
Please, squash migration commits before merging for cleaner commits history

**Comment by pedrobaeza** [comment]
:+1:

**Comment by abdufarag** [comment]
Error while importing module 'account_invoice_refund_link': u'"Error while validating constraint\n\nField `refund_reason` does not exist\n\nError context:\nView `Account invoice (customer) | add Refunds details (form)`\n[view_id: 813, xml_id: n/a, model: account.invoice, parent_id: 502]\nNone" while parsing /tmp/tmpxubF2O/account_invoice_refund_link/views/account_invoice_view.xml:8, near\n<record id="view_customer_account_invoice_add_refunds_details_form" model="ir.ui.view">\n <field name="name">Account invoice (customer) | add Refunds details (form)</field>\n <field name="model">account.invoice</field>\n <field name="inherit_id" ref="account.invoice_form"/>\n <field name="arch" type="xml">\n <notebook position="inside">\n <page name="refunds" string="Refunds">\n <group attrs="{\'invisible\':[(\'type\', \'not in\', [\'in_refund\', \'out_refund\'])]}">\n <separator string="Description" colspan="4"/>\n <field name="refund_reason" nolabel="1"/>\n <separator string="Original invoice" colspan="4"/>\n <field name="origin_invoice_ids" nolabel="1">\n <tree decoration-info="state == \'draft\'" decoration-muted="state == \'cancel\'" string="Original invoice">\n <field name="partner_id" string="Customer"/>\n <field name="date_invoice" string="Bill date"/>\n <field name="date_due" string="Due date"/>\n <field name="number" string="Number"/>\n <field name="reference" string="Reference"/>\n <field name="origin" string="Source"/>\n <field name="amount_total_signed" string="Total"/>\n <field name="residual_signed" string="Due"/>\n <field name="state" string="Status"/>\n </tree>\n </field>\n </group>\n <group attrs="{\'invisible\':[(\'type\', \'not in\', [\'in_invoice\', \'out_invoice\'])]}">\n <separator string="Refund invoices" colspan="4"/>\n <field name="refund_invoice_ids" nolabel="1">\n <tree decoration-info="state == \'draft\'" decoration-muted="state == \'cancel\'" string="Refund invoices">\n <field name="partner_id" string="Customer"/>\n <field name="date_invoice" string="Bill date"/>\n <field name="number" string="Number"/>\n <field name="reference" string="Reference"/>\n <field name="origin" string="Source"/>\n <field name="refund_reason" string="Refund reason"/>\n <field name="amount_total" string="Total"/>\n <field name="state" string="Status"/>\n </tree>\n </field>\n </group>\n </page>\n </notebook>\n </field>\n</record>'
